src/candid_rpc.rs

```diff
@@ -0,0 +1,174 @@
+use std::str::FromStr;
+
+use async_trait::async_trait;
+use cketh_common::{
+    eth_rpc::{
+        into_nat, Block, FeeHistory, GetLogsParam, Hash, HttpOutcallError, JsonRpcReply, LogEntry,
+        ProviderError, RpcError, SendRawTransactionResult, ValidationError,
+    },
+    eth_rpc_client::{
+        providers::{RpcApi, RpcNodeProvider},
+        requests::GetTransactionCountParams,
+        EthRpcClient as CkEthRpcClient, MultiCallError, RpcTransport,
+    },
+    lifecycle::EthereumNetwork,
+};
+use serde::de::DeserializeOwned;
+
+use crate::*;
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+struct CanisterTransport;
+
+#[cfg_attr(target_arch = "wasm32", async_trait(?Send))]
+#[cfg_attr(not(target_arch = "wasm32"), async_trait)]
+impl RpcTransport for CanisterTransport {
+    fn get_subnet_size() -> u32 {
+        METADATA.with(|m| m.borrow().get().nodes_in_subnet)
+    }
+
+    fn resolve_api(provider: RpcNodeProvider) -> Result<RpcApi, ProviderError> {
+        // TODO: https://github.com/internet-computer-protocol/ic-eth-rpc/issues/73
+        Ok(provider.api())
+    }
+
+    async fn call_json_rpc<T: DeserializeOwned>(
+        provider: RpcNodeProvider,
+        json: &str,
+        max_response_bytes: u64,
+    ) -> Result<T, RpcError> {
+        let response = do_http_request(
+            ic_cdk::caller(),
+            ResolvedSource::Api(Self::resolve_api(provider)?),
+            json,
+            max_response_bytes,
+        )
+        .await
+        .unwrap();
+        let status = get_http_response_status(response.status.clone());
+        let body = get_http_response_body(response)?;
+        let json: JsonRpcReply<T> = serde_json::from_str(&body).unwrap_or_else(|e| {
+            Err(HttpOutcallError::InvalidHttpJsonRpcResponse {
+                status,
+                body,
+                parsing_error: Some(format!("JSON response parse error: {e}")),
+            })
+        })?;
+        json.result.into()
+    }
+}
+
+fn get_rpc_client(source: CandidRpcSource) -> RpcResult<CkEthRpcClient<CanisterTransport>> {
+    fn validate_providers<T>(opt_vec: Option<Vec<T>>) -> RpcResult<Option<Vec<T>>> {
+        Ok(match opt_vec {
+            Some(v) if v.is_empty() => Err(ProviderError::ProviderNotFound)?,
+            opt => opt,
+        })
+    }
+    if !is_rpc_allowed(&ic_cdk::caller()) {
+        // inc_metric!(eth_*_err_no_permission);
+        return Err(ProviderError::NoPermission.into());
+    }
+    Ok(match source {
+        CandidRpcSource::EthMainnet(service) => CkEthRpcClient::new(
+            EthereumNetwork::Ethereum,
+            validate_providers(Some(vec![service.unwrap_or(
+                cketh_common::eth_rpc_client::providers::EthereumProvider::Ankr,
```

Comment:
Is Ankr used here as a default? A comment would help here.

Reply:
Good point; I moved these to `constants.rs` with a comment explaining when the canister uses the defaults.